### PR TITLE
Retry blob upload indefinitely on rejection

### PR DESCRIPTION
**Pull request by dannimad**
Given the recent addition of abort signal, user have more control over when to cancel a stale or failing blob upload. We can retry indefinitely up to user abort signal adding resiliency to uploadBlob method. 

[AB#5392](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/5392)

### CODE REVIEW

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 1]
```diff
@@ -568,23 +569,23 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
...
-			// we are probably not connected to storage but start another upload request in case we are
-			entry.uploadP = this.retryThrottler
-				.getDelay()
-				.then(async () => this.uploadBlob(localId, entry.blob));
-			return entry.uploadP;
-		} else {
-			entry.handleP.reject(error);
-			throw error;
-		}
+		this.mc.logger.sendTelemetryEvent({
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/blobManager.ts:579]: can we add tests for this?

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 2]
```diff
@@ -568,23 +569,23 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
 		return response;
 	}
 
-	private async onUploadReject(localId: string, error: any) {
+	private async onUploadReject(localId: string, error: any, count: number) {
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/blobManager.ts:572]: can we simplify this, maybe use driver-utils runWithRetry instead?

Comment (anthony-murphy) [reply packages/runtime/container-runtime/src/blobManager.ts:572]: maybe something like this:
```
	private async uploadBlob(
		localId: string,
		blob: ArrayBufferLike,
	): Promise<ICreateBlobResponse | void> {
		return runWithRetry(
			async () => this.getStorage().createBlob(blob),
			"createBlob",
			this.mc.logger,
			{
				cancel: this.pendingBlobs.get(localId)?.abortSignal,
			},
		).then((response) => this.onUploadResolve(localId, response));
	}
```

Comment (dannimad) [reply packages/runtime/container-runtime/src/blobManager.ts:572]: I like using runWithRetry method. There is some cleanup to do in case of ending up rejecting but I can add that to the method.

**Comment by dannimad** [comment]
runWithRetry is now catching this previously ignored error in some stashed blob tests:
``` 
Uncaught Error: Real storage calls not allowed in Unattached container
      at BlobOnlyStorage.verifyStorage (/workspaces/FluidFramework/packages/loader/container-loader/src/containerStorageAdapter.ts:200:10)
      at BlobOnlyStorage.createBlob (/workspaces/FluidFramework/packages/loader/container-loader/src/containerStorageAdapter.ts:191:15)
      at ContainerStorageAdapter.createBlob (/workspaces/FluidFramework/packages/loader/container-loader/src/containerStorageAdapter.ts:176:31)
      at /workspaces/FluidFramework/packages/runtime/container-runtime/src/blobManager.ts:498:34
      at runWithRetry (/workspaces/FluidFramework/packages/loader/driver-utils/src/runWithRetry.ts:61:19)
      at BlobManager.uploadBlob (/workspaces/FluidFramework/packages/runtime/container-runtime/src/blobManager.ts:497:22)
      at BlobManager.createBlob (/workspaces/FluidFramework/packages/runtime/container-runtime/src/blobManager.ts:473:18)
      at ContainerRuntime.uploadBlob (/workspaces/FluidFramework/packages/runtime/container-runtime/src/containerRuntime.ts:3389:27)
      at RemoteFluidDataStoreContext.uploadBlob (/workspaces/FluidFramework/packages/runtime/container-runtime/src/dataStoreContext.ts:943:32)
      at RuntimeWithRequestHandler.uploadBlob (/workspaces/FluidFramework/packages/runtime/datastore/src/dataStoreRuntime.ts:572:32)
```
I believe this error can continue to be ignored (still looking) but UsageError by default canRetry value is false.

**Comment by anthony-murphy** [comment]
> runWithRetry is now catching this previously ignored error in some stashed blob tests:
> 
> ```
> Uncaught Error: Real storage calls not allowed in Unattached container
>       at BlobOnlyStorage.verifyStorage (/workspaces/FluidFramework/packages/loader/container-loader/src/containerStorageAdapter.ts:200:10)
>       at BlobOnlyStorage.createBlob (/workspaces/FluidFramework/packages/loader/container-loader/src/containerStorageAdapter.ts:191:15)
>       at ContainerStorageAdapter.createBlob (/workspaces/FluidFramework/packages/loader/container-loader/src/containerStorageAdapter.ts:176:31)
>       at /workspaces/FluidFramework/packages/runtime/container-runtime/src/blobManager.ts:498:34
>       at runWithRetry (/workspaces/FluidFramework/packages/loader/driver-utils/src/runWithRetry.ts:61:19)
>       at BlobManager.uploadBlob (/workspaces/FluidFramework/packages/runtime/container-runtime/src/blobManager.ts:497:22)
>       at BlobManager.createBlob (/workspaces/FluidFramework/packages/runtime/container-runtime/src/blobManager.ts:473:18)
>       at ContainerRuntime.uploadBlob (/workspaces/FluidFramework/packages/runtime/container-runtime/src/containerRuntime.ts:3389:27)
>       at RemoteFluidDataStoreContext.uploadBlob (/workspaces/FluidFramework/packages/runtime/container-runtime/src/dataStoreContext.ts:943:32)
>       at RuntimeWithRequestHandler.uploadBlob (/workspaces/FluidFramework/packages/runtime/datastore/src/dataStoreRuntime.ts:572:32)
> ```
> 
> I believe this error can continue to be ignored (still looking) but UsageError by default canRetry value is false.

this looks like a real error to me. if the test wants to use blobs while detached it should use detached blob storage.

File: packages/loader/driver-utils/src/runWithRetry.ts [hunk 3]
```diff
@@ -48,6 +48,7 @@ export async function runWithRetry<T>(
 	fetchCallName: string,
 	logger: ITelemetryLoggerExt,
 	progress: IProgress,
+	beforeErrorThrow?: () => void,
```
Comment (anthony-murphy) [inline packages/loader/driver-utils/src/runWithRetry.ts:51]: this shouldn't be necessary, just do the same thing in a .catch of the promise returned by the function

File: packages/loader/container-loader/src/retriableDocumentStorageService.ts [hunk 4]
```diff
@@ -23,13 +23,25 @@ import { runWithRetry } from "@fluidframework/driver-utils";
 
 export class RetriableDocumentStorageService implements IDocumentStorageService, IDisposable {
 	private _disposed = false;
+	private internalStorageService: IDocumentStorageService | undefined;
+	private readonly storageServiceDeferred: Deferred<void> = new Deferred();
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/retriableDocumentStorageService.ts:27]: i don't think you need this. the promise, and the value should be enough

File: packages/loader/container-loader/src/retriableDocumentStorageService.ts [hunk 5]
```diff
@@ -39,22 +51,35 @@ export class RetriableDocumentStorageService implements IDocumentStorageService,
...
+			return this.internalStorageService.repositoryUrl;
+		}
+		throw new Error("storage service not yet instantiated");
 	}
 
 	public async getSnapshotTree(
 		version?: IVersion,
 		scenarioName?: string,
 	): Promise<ISnapshotTree | null> {
+		if (this.internalStorageService === undefined) {
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/retriableDocumentStorageService.ts:64]: remove this

File: packages/loader/container-loader/src/retriableDocumentStorageService.ts [hunk 6]
```diff
@@ -39,22 +51,35 @@ export class RetriableDocumentStorageService implements IDocumentStorageService,
...
+			await this.storageServiceDeferred.promise;
+		}
 		return this.runWithRetry(
-			async () => this.internalStorageService.getSnapshotTree(version, scenarioName),
+			async () =>
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/retriableDocumentStorageService.ts:68]:
```suggestion
			async () =>
				this.internalStorageServiceP.then(async (s)=>s.getSnapshotTree(
					version,
					scenarioName,
				)),```

Comment (anthony-murphy) [reply packages/loader/container-loader/src/retriableDocumentStorageService.ts:68]: do this for all the methods and remove the deferred

Comment (anthony-murphy) [reply packages/loader/container-loader/src/retriableDocumentStorageService.ts:68]: also, casting is bad, it should almost never be done, and when done it should be limit to as a few places as possible

File: packages/loader/container-loader/src/retriableDocumentStorageService.ts [hunk 7]
```diff
@@ -95,26 +108,32 @@ export class RetriableDocumentStorageService implements IDocumentStorageService,
...
+			return this.internalStorageServiceP.then(async (s) =>
+				s.uploadSummaryWithContext(summary, context),
+			);
 		}
 
 		// Creation flow with attachment blobs - need to do retries!
 		return this.runWithRetry(
-			async () => this.internalStorageService.uploadSummaryWithContext(summary, context),
+			async () =>
+				(this.internalStorageService as IDocumentStorageService).uploadSummaryWithContext(
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/retriableDocumentStorageService.ts:119]: we shouldn't be casting here. just use the promise like above

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 8]
```diff
@@ -493,15 +494,21 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
...
+			"createBlob",
+			this.mc.logger,
+			{
+				cancel: this.pendingBlobs.get(localId)?.abortSignal,
+			},
+		)
+			.then((response) => this.onUploadResolve(localId, response))
+			.catch((err) => {
+				this.pendingBlobs.get(localId)?.handleP.reject(err);
+				if (!this.pendingBlobs.get(localId)?.opsent) {
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/blobManager.ts:508]: if the op has been sent and so we don't delete the pending blob, how does the upload get restarted?

File: packages/loader/container-loader/src/retriableDocumentStorageService.ts (hunk 7 above)
Comment (dannimad) [reply packages/loader/container-loader/src/retriableDocumentStorageService.ts:119]: thanks, I missed this one.

File: packages/runtime/container-runtime/src/blobManager.ts (hunk 8 above)
Comment (dannimad) [reply packages/runtime/container-runtime/src/blobManager.ts:508]: The only way the op has been sent at this point is because it's being stashed. We want to avoid messing with pendingBlobs array during that process.

File: packages/loader/container-loader/src/retriableDocumentStorageService.ts [hunk 9]
```diff
@@ -23,13 +23,23 @@ import { runWithRetry } from "@fluidframework/driver-utils";
...
 export class RetriableDocumentStorageService implements IDocumentStorageService, IDisposable {
 	private _disposed = false;
+	private internalStorageService: IDocumentStorageService | undefined;
 	constructor(
-		private readonly internalStorageService: IDocumentStorageService,
+		private readonly internalStorageServiceP: Promise<IDocumentStorageService>,
 		private readonly logger: ITelemetryLoggerExt,
-	) {}
+	) {
+		this.initialize().catch(() => {});
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/retriableDocumentStorageService.ts:31]:
```suggestion
		this.internalStorageServiceP.then((s)=>this.internalStorageService =s).catch(() => {});
```

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 10]
```diff
@@ -493,15 +466,29 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
...
-			async (err) => this.onUploadReject(localId, err),
-		);
+			"createBlob",
+			this.mc.logger,
+			{
+				cancel: this.pendingBlobs.get(localId)?.abortSignal,
+			},
+		)
+			.then((response) => this.onUploadResolve(localId, response))
+			.catch((err) => {
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/blobManager.ts:478]: i think this code should go on the inside of the retry loop, not outside it. something like 

```
async () => {
		try{
   				 return await this.getStorage().createBlob(blob);
		}catch(error){
				const entry = this.pendingBlobs.get(localId);
				assert(
					!!entry,
					0x387 /* Must have pending blob entry for blob which failed to upload */,
				);
				if (entry.opsent && !canRetryOnError(error)) {
					/// this is the important part
					throw isFluidError(e) ? e.addTelemetryProperties({canRetry: true}) : wrapError(e, ()=>createErrorObjectWithProps("createBlob");
				}
				entry.handleP.reject(err);
				this.deletePendingBlob(localId);
				throw error;
				
		}
}
```

Comment (anthony-murphy) [reply packages/runtime/container-runtime/src/blobManager.ts:478]: the problem with the existing code is there is no retry delay if the error is not retriable, so this change makes all errors retriable when we have to succeed.

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 11]
```diff
@@ -467,28 +473,51 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
...
+					return await this.getStorage().createBlob(blob);
+				} catch (error) {
+					const entry = this.pendingBlobs.get(localId);
+					assert(
+						!!entry,
+						0x387 /* Must have pending blob entry for blob which failed to upload */,
+					);
+					if (entry.opsent && !canRetryOnError(error)) {
+						if (isFluidError(error)) {
+							error.addTelemetryProperties({ forceRetry: true });
```
Comment (dannimad) [inline packages/runtime/container-runtime/src/blobManager.ts:487]: addTelemetryProperties can't overwrite them. I thought about a couple of ways to tackle it and this one avoids adding a lot of unnecessary changes in the error types.

File: packages/runtime/container-runtime/src/test/blobManager.spec.ts [hunk 12]
```diff
@@ -172,15 +173,23 @@ export class MockRuntime
...
 		} else {
-			const err: any = new Error("fake error");
-			err.canRetry = canRetry;
-			err.retryAfterSeconds = retryAfterSeconds;
+			const err = createGenericNetworkError(
+				"fake driver error",
+				// TODO: https://dev.azure.com/fluidframework/internal/_workitems/edit/5913
+				{ canRetry, retryAfterMs },
+				{
+					// To avoid using the default retry time added when we don't have retryAfterMs
```
Comment (dannimad) [inline packages/runtime/container-runtime/src/test/blobManager.spec.ts:187]: I created an item to track a but I saw here where we are losing the retryAfterMs property when creating a GenericNetworkError

File: packages/loader/driver-utils/src/network.ts [hunk 13]
```diff
@@ -193,7 +193,8 @@ export function createGenericNetworkError(
  * I.e. asserts or unexpected exceptions in our code result in container failure.
  * @param error - The error to inspect for ability to retry
  */
-export const canRetryOnError = (error: any): boolean => error?.canRetry === true;
+export const canRetryOnError = (error: any): boolean =>
+	error?.canRetry === true || error?.forceRetry === true;
```
Comment (anthony-murphy) [inline packages/loader/driver-utils/src/network.ts:197]: still reading, but i'm prefer it we didn't have to modify this

File: packages/runtime/container-runtime/src/test/blobManager.spec.ts [hunk 14]
```diff
@@ -172,13 +173,24 @@ export class MockRuntime
...
 		} else {
-			this.processBlobsP.reject(new Error("fake error"));
+			const err = createGenericNetworkError(
+				"fake driver error",
+				// TODO: https://dev.azure.com/fluidframework/internal/_workitems/edit/5913
+				{ canRetry, retryAfterMs },
+				{
+					// To avoid using the default retry time added when we don't have retryAfterMs
+					// in the error properties.
+					retryAfterSeconds: retryAfterMs,
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/test/blobManager.spec.ts:189]: this looks wrong. seconds !== ms

File: packages/loader/driver-utils/src/network.ts (hunk 13 above)
Comment (dannimad) [reply packages/loader/driver-utils/src/network.ts:197]: We could modify the RunWithRetry method instead of adding forceRetry to canRetryOnError

File: packages/runtime/container-runtime/src/test/blobManager.spec.ts [hunk 15]
```diff
@@ -448,21 +468,38 @@ describe("BlobManager", () => {
...
 		assert.strictEqual(summaryData.redirectTable, undefined);
 	});
 
+	it("upload fails and retries for retriable errors", async () => {
+		await runtime.attach();
+		await runtime.connect();
+		const handleP = runtime.createBlob(IsoBuffer.from("blob", "utf8"));
+		await runtime.processBlobs(false, true, 0);
+		// wait till next retry
+		await new Promise<void>((resolve) => setTimeout(resolve, 1));
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/test/blobManager.spec.ts:485]: what is this doing specifically? is it really necessary? does the test pass without? just waiting like this isn't a stable or consistent way to ensure something happens. you need to wait on something specific.

File: packages/runtime/container-runtime/src/test/blobManager.spec.ts (hunk 14 above)
Comment (dannimad) [reply packages/runtime/container-runtime/src/test/blobManager.spec.ts:189]: This is a workaround, it shouldn't be there in the first place and it's only for these tests to avoid the timeout. I open a task to fix it

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 16]
```diff
@@ -493,14 +472,51 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
...
+									`uploadBlob error`,
+									{ canRetry: true },
+									{
+										driverVersion: undefined,
+									},
+								),
+							);
+						}
+					}
+					if (canRetryOnError(error)) {
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/blobManager.ts:501]: this should just throw, and the delete should happen in a .catch outside of the retry

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 17]
```diff
@@ -493,14 +472,51 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
...
-			{ eventName: "createBlob" },
-			async () => this.getStorage().createBlob(blob),
-			{ end: true, cancel: this.runtime.connected ? "error" : "generic" },
+			{
+				cancel: this.pendingBlobs.get(localId)?.abortSignal,
+			},
 		).then(
-			(response) => this.onUploadResolve(localId, response),
-			async (err) => this.onUploadReject(localId, err),
+			(response) => response && this.onUploadResolve(localId, response),
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/blobManager.ts:518]: when is response undefined?

File: packages/runtime/container-runtime/src/blobManager.ts (hunk 11 above)
Comment (anthony-murphy) [reply packages/runtime/container-runtime/src/blobManager.ts:487]: the thing the wrap error code below will do the right thing in all cases. i think we can just remove the if, and only do that

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 18]
```diff
@@ -493,14 +472,51 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
...
+						!!entry,
+						0x387 /* Must have pending blob entry for blob which failed to upload */,
+					);
+					if (entry.opsent && !canRetryOnError(error)) {
+						if (isFluidError(error)) {
+							error.addTelemetryProperties({ forceRetry: true });
+							throw error;
+						} else {
+							throw wrapError(error, () =>
+								createGenericNetworkError(
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/blobManager.ts:491]: I don't think createGenericNetworkError is really appropriate here, as we don't know anything about the inner error, so we don't know if it is network or not. i think just creating a new logging error would be enough, and wrap error will take most the properties from the inner error.

Comment (anthony-murphy) [reply packages/runtime/container-runtime/src/blobManager.ts:491]: the biggest problem is that this approach overwrites the errorType, using logging error will preserve it.

File: packages/runtime/container-runtime/src/test/blobManager.spec.ts (hunk 15 above)
Comment (dannimad) [reply packages/runtime/container-runtime/src/test/blobManager.spec.ts:485]: Without this promise, the following processBlobs will fail to see the 2nd uploadBlob call made by runWithRetry. Specifically, it's waiting for `await delay(retryAfterMs)` line in RunWithRetry. Setting the delay time to 0 in the previous processBlobs call should be ok to ensure the correct order of things, but maybe there is a better way to wait for it.

**Review by anthony-murphy — APPROVED** [review]
